core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2242,12 +2242,37 @@ object ZIO {
    * produced effects in parallel, discarding the results.
    *
    * For a sequential version of this method, see `foreach_`.
+   *
+   * Optimized to avoid keeping full tree of effects, so that method could be
+   * able to handle large input sequences.
+   * Behaves almost like this code:
+   *
+   * {{{
+   * as.foldLeft(ZIO.unit) { (acc, a) => acc.zipParLeft(f(a)) }
+   * }}}
+   *
+   * Additionally, interrupts all effects on any failure.
    */
-  def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, Any]): ZIO[R, E, Unit] =
-    as.foldLeft(unit: ZIO[R, E, Unit]) { (acc, a) =>
-        acc.zipParLeft(f(a))
-      }
-      .refailWithTrace
+  def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, Any]): ZIO[R, E, Unit] = {
+    val size = as.size
+    for {
+      parentId <- ZIO.fiberId
+      result   <- Promise.make[E, Unit]
+      succeed  <- Ref.make(0)
+      _ <- ZIO.traverse_(as) {
+            f(_)
+              .foldM(result.fail, _ => {
```

Comment:
Currently defects in a fiber will not cause the other fibers to be terminated. So let's do this:
```suggestion
              .foldCauseM(result.halt, _ => {
```